### PR TITLE
feature(whl): add DREX algorithm

### PR DESCRIPTION
**Pull request by kxzxvbk**
## Description
Implementation of DREX for IRL.
Reference: https://arxiv.org/abs/1907.03976

## TODO
1. combine the serial entry with trex.
2. add on policy version.
3. add test files

## Check List

- [x] merge the latest version source branch/repo, and resolve all the conflicts
- [x] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/entry/application_entry_drex_collect_data.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+from ding.config import compile_config, read_config
+from ding.worker import EpisodeSerialCollector, create_buffer, BaseLearner
+from ding.envs import create_env_manager, get_vec_env_setting
+from ding.policy import create_policy, bc
+from ding.torch_utils import to_device
+from ding.utils import set_pkg_seed
+from ding.utils.data import default_collate, offline_data_save_type
+
+
+def collect_episodic_demo_data_for_drex(
```
Comment (Will-Nie) [inline ding/entry/application_entry_drex_collect_data.py:21]: add test for new added function (see trex for reference)

File: ding/entry/serial_entry_drex.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                break
+            # update train_data reward
+            reward_model.estimate(train_data)
+            learner.train(train_data, collector.envstep)
+            if learner.policy.get_attribute('priority'):
+                replay_buffer.update(learner.priority_info)
+
+    # Learner's after_run hook.
+    learner.call_hook('after_run')
+    return policy
```
[the remark this answers is not in the document]
Comment (PaParaZz1) [reply ding/entry/serial_entry_drex.py:136]: > can we combine this serial entry with trex ? (e.g. change serial_pipeline_reward_model_trex into serial_pipeline_reward_model_perferecnce_based_irl)

+1

File: ding/model/template/bc.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import torch
+import torch.nn as nn
+
+from ding.torch_utils import get_lstm
+from ding.utils import MODEL_REGISTRY, SequenceType, squeeze
+from ..common import FCEncoder, ConvEncoder, DiscreteHead, DuelingHead, MultiHead
+
+
+@MODEL_REGISTRY.register('bc')
+class BC(nn.Module):
```
Comment (PaParaZz1) [inline ding/model/template/bc.py:11]: `DiscreteBehaviourCloning` can be better

File: ding/policy/bc.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    config = dict(
+        type='fc',
+        on_policy=False,
+        collect=dict(
+            # (int) Only one of [n_sample, n_step, n_episode] shoule be set
+            #n_sample=8,  # collect 8 samples and put them in collector
+            # (int) Cut trajectories into pieces with length "unroll_len".
+            unroll_len=1,
+        ),
+    )
```
Comment (PaParaZz1) [inline ding/policy/bc.py:35]: add all the common hyper-parameters in config, such as `learning_rate`

File: ding/policy/bc.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        r"""
+        Overview:
+            Collect mode init method. Called by ``self.__init__``.
+            Init traj and unroll length, collect model.
+            Enable the eps_greedy_sample
+        """
+        self._unroll_len = self._cfg.collect.unroll_len
+        #self._gamma = self._cfg.discount_factor  # necessary for parallel
+        #self._nstep = self._cfg.nstep  # necessary for parallel
+        self._collect_model = model_wrap(self._model, wrapper_name='eps_greedy_sample')
```
Comment (PaParaZz1) [inline ding/policy/bc.py:129]: why bc needs `collect` part

File: dizoo/box2d/lunarlander/config/lunarlander_drex_dqn_config.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    env=dict(
+        type='lunarlander',
+        import_names=['dizoo.box2d.lunarlander.envs.lunarlander_env'],
+    ),
+    env_manager=dict(type='base'),
+    policy=dict(type='dqn'),
+)
+lunarlander_drex_dqn_create_config = EasyDict(lunarlander_drex_dqn_create_config)
+create_config = lunarlander_drex_dqn_create_config
+main_config
```
Comment (PaParaZz1) [inline dizoo/box2d/lunarlander/config/lunarlander_drex_dqn_config.py:94]: unfinished `main_config`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/bc.py (hunk 5 above)
Comment (kxzxvbk) [reply ding/policy/bc.py:133]: In the latter part, DREX will inject noise into the BC policy (using epsilon greedy model wrapper). So I use the collect mode to do that.

File: dizoo/atari/config/serial/spaceinvaders/spaceinvaders_drex_dqn_config.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
+from easydict import EasyDict
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/spaceinvaders/spaceinvaders_drex_dqn_config.py:1]: modify config according to our new defined config specification

File: dizoo/atari/config/serial/spaceinvaders/spaceinvaders_drex_dqn_config.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    ),
+    reward_model=dict(
+        type='drex',
+        algo_for_model='dqn',
+        env_id='SpaceInvadersNoFrameskip-v4',
+        min_snippet_length=30,
+        max_snippet_length=100,
+        learning_rate=1e-5,
+        update_per_collect=1,
+        expert_model_path='abs model path',
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/spaceinvaders/spaceinvaders_drex_dqn_config.py:22]: what is the relationship between `expert_model_path` and `bc_path`

File: dizoo/atari/config/serial/spaceinvaders/spaceinvaders_drex_dqn_config.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        learn=dict(
+            update_per_collect=10,
+            batch_size=32,
+            learning_rate=0.0001,
+            target_update_freq=500,
+        ),
+        collect=dict(
+            n_sample=100,
+            collector=dict(
+                get_train_sample=False,
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/spaceinvaders/spaceinvaders_drex_dqn_config.py:50]: why add this argument, if you use `SampleSerialCollector`

File: ding/reward_model/drex_reward_model.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+            traj_j = self.created_data[bj][tj][tj_start:tj_start + rand_length[n]:2]
+
+            max_traj_length = max(max_traj_length, len(traj_i), len(traj_j))
+            label = int(bi <= bj)
+            self.training_obs.append((traj_i, traj_j))
+            self.training_labels.append(label)
+        self._logger.info(("maximum traj length: {}".format(max_traj_length)))
+        return self.training_obs, self.training_labels
+
+    def train(self):
```
Comment (PaParaZz1) [inline ding/reward_model/drex_reward_model.py:141]: what is difference of train method between trex and drex, I think there should be no difference between them. As for implementation like `pred_data` method, you can also add this method in trex to unify them.

File: ding/reward_model/drex_reward_model.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+        with open(self.cfg.reward_model.offline_data_path + '/created_data_returns.pkl', 'rb') as f:
+            self.created_data_returns = pickle.load(f)
+        with open(self.cfg.reward_model.offline_data_path + '/suboptimal_data.pkl', 'rb') as f:
+            self.pre_expert_data = pickle.load(f)
+        self.create_training_data()
+
+        self._logger.info("num_training_obs: {}".format(len(self.training_obs)))
+        self._logger.info("num_labels: {}".format(len(self.training_labels)))
+
+    def create_training_data(self):
```
Comment (PaParaZz1) [inline ding/reward_model/drex_reward_model.py:69]: what is difference of `create_training_data` method between trex and drex

File: ding/policy/bc.py (hunk 5 above)
Comment (PaParaZz1) [reply ding/policy/bc.py:133]: add some comments to explain this

File: ding/policy/bc.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+from ding.policy import Policy
+from ding.model import model_wrap
+from ding.torch_utils import to_device
+from ding.utils import EasyTimer
+from ding.utils.data import default_collate, default_decollate
+from ding.rl_utils import q_nstep_td_data, q_nstep_sql_td_error, get_nstep_return_data, get_train_sample
+from ding.worker.collector.interaction_serial_evaluator import InteractionSerialEvaluator
+from ding.utils import POLICY_REGISTRY
+
+
```
Comment (PaParaZz1) [inline ding/policy/bc.py:19]: merge some functions of `policy/il.py` info this `policy/bc.py`, you can ignore some parts code related to gfootball. As the result, the modified code can pass unittest in `ding/entry/tests/test_serial_pipeline_il.py`.

File: ding/entry/serial_pipeline_reward_model_reward_model_preference_based_irl_onpolicy.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
+from typing import Union, Optional, List, Any, Tuple
```
Comment (PaParaZz1) [inline ding/entry/serial_pipeline_reward_model_reward_model_preference_based_irl_onpolicy.py:1]: add unittest for entries related to drex

File: ding/entry/serial_pipeline_reward_model_preference_based_irl.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+from ding.reward_model import create_reward_model
+from ding.utils import set_pkg_seed
+
+
+def serial_pipeline_reward_model_reward_model_preference_based_irl(
+        input_cfg: Union[str, Tuple[dict, dict]],
+        seed: int = 0,
+        env_setting: Optional[List[Any]] = None,
+        model: Optional[torch.nn.Module] = None,
+        max_iterations: Optional[int] = int(1e10),
```
Comment (PaParaZz1) [inline ding/entry/serial_pipeline_reward_model_preference_based_irl.py:23]: modify `max_iterations` like other entry

File: ding/entry/application_entry_drex_collect_data.py [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+    for _, data in enumerate(validation_set):
+        x, y = {'obs': data['obs'].to(device).squeeze(0)}, data['action'].squeeze(-1)
+        y_pred = policy.forward(x, eps=-1)['obs']['action']
+        tot += y_pred.shape[0]
+        tot_acc += (y_pred == y).sum().item()
+    acc = tot_acc / tot
+    return acc
+
+
+def train_bc(cfg, pre_expert_data=None, max_iterations=6000):
```
Comment (PaParaZz1) [inline ding/entry/application_entry_drex_collect_data.py:104]: merge `train_bc` and `evaluate_bc` functions into `serial_pipeline_il` (or you can rename it to `serial_pipeline_bc`)

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**